app/repositories/LockRepository.scala

```diff
@@ -0,0 +1,163 @@
+    implicit val format: OFormat[JsonDataEntry] = Json.format[JsonDataEntry]
+  }
+
+
+  override lazy val indexes: Seq[Index] = Seq(
+    Index(key = Seq("psaId" -> Ascending), name = Some("psaId_Index"), unique = true),
+    Index(key = Seq("srn" -> Ascending), name = Some("srn_Index"), unique = true),
+    Index(key = Seq("expireAt" -> Ascending), name = Some("dataExpiry"), options = BSONDocument("expireAfterSeconds" -> 0))
+  )
+
+  override def ensureIndexes(implicit ec: ExecutionContext): Future[Seq[Boolean]] = {
+    Future.sequence(indexes.map(collection.indexesManager.ensure(_)))
+  }
+
+  override def releaseLock(lock: SchemeVariance): Future[Unit] = collection.findAndRemove(byLock(lock.psaId, lock.srn)).map(_ => ())
+
+  override def releaseLockByPSA(psaId: String): Future[Unit] = collection.findAndRemove(byPsaId(psaId)).map(_ => ())
+
+  override def releaseLockBySRN(srn: String): Future[Unit] = collection.findAndRemove(bySrn(srn)).map(_ => ())
+
+  override def getExistingLock(lock: SchemeVariance): Future[Option[SchemeVariance]] = collection.find(byLock(lock.psaId, lock.srn)).one[SchemeVariance]
+
+  override def getExistingLockByPSA(psaId: String): Future[Option[SchemeVariance]] = collection.find(byPsaId(psaId)).one[SchemeVariance]
+
+  override def getExistingLockBySRN(srn: String): Future[Option[SchemeVariance]] = collection.find(bySrn(srn)).one[SchemeVariance]
+
+  override def list: Future[List[SchemeVariance]] = {
+    val arbitraryLimit = 10000
+    collection.find(Json.obj())
+      .cursor[SchemeVariance]()
+      .collect[List](arbitraryLimit, Cursor.FailOnError())
+  }
+
+  override def replaceLock(newLock: SchemeVariance): Future[Boolean] = {
+    collection.update(
+      byLock(newLock.psaId,newLock.srn), modifier(newLock), upsert = true).map {
+      lastError =>
+        lastError.writeErrors.isEmpty
+    } recoverWith {
+      case e: LastError if e.code == documentExistsErrorCode => {
+        getExistingLock(newLock).map{
+          case Some(existingLock) => existingLock.psaId==newLock.psaId && existingLock.srn==newLock.srn
+          case None => throw new Exception(s"Expected SchemeVariance to be locked, but no lock was found with psaId: ${newLock.psaId} and srn: ${newLock.srn}")
+        }
+      }
+    }
+  }
+
+  override def lock(newLock: SchemeVariance): Future[Lock] = {
+
+    collection.update(byLock(newLock.psaId, newLock.srn), modifier(newLock), upsert = true)
+      .map[Lock](_ => VarianceLock) recoverWith {
+      case e: LastError if e.code == documentExistsErrorCode => {
+        for{
+          psaLock <- getExistingLockByPSA(newLock.psaId)
+          srnLock <- getExistingLockBySRN(newLock.srn)
+        } yield {
+          (psaLock, srnLock) match {
+            case (Some(_), None) => PsaLock
+            case (None, Some(_)) => SchemeLock
+            case (Some(id), Some(srn)) => if(id==newLock.psaId && srn == newLock.srn) VarianceLock else BothLock
```

Comment:
Comparing unrelated types. Should be id.psaId==... and srn.srn. Also local name should not be Some(id) as it is not id but case class instance